File: Languages/python/algorithms-dsa/patterns/sliding-window/best_time_buy_sell.py
"""
LeetCode 121: Best Time to Buy and Sell Stock
Given an array prices where prices[i] is the price on the ith day,
find the maximum profit from one transaction.

Time: O(n), Space: O(1)
"""


def max_profit(prices: list[int]) -> int:
    min_price = float("inf")
    profit = 0

    for price in prices:
        min_price = min(min_price, price)
        profit = max(profit, price - min_price)

    return profit


if __name__ == "__main__":
    assert max_profit([7, 1, 5, 3, 6, 4]) == 5
    assert max_profit([7, 6, 4, 3, 1]) == 0
    assert max_profit([1, 2]) == 1
    assert max_profit([2, 1, 4]) == 3
    assert max_profit([]) == 0

    print("All tests passed!")

File: Languages/cpp/algorithms-dsa/patterns/sliding-window/best_time_buy_sell.cpp
/**
 * Pattern: Sliding Window
 * Problem: Best Time to Buy and Sell Stock (LeetCode 121)
 *
 * Find max profit from one buy and one sell.
 * Time: O(n), Space: O(1)
 */
#include <iostream>
#include <vector>
#include <algorithm>
#include <cassert>
using namespace std;

int maxProfit(vector<int>& prices) {
    int minPrice = prices[0];
    int maxProfit = 0;
    for (int i = 1; i < (int)prices.size(); i++) {
        maxProfit = max(maxProfit, prices[i] - minPrice);
        minPrice = min(minPrice, prices[i]);
    }
    return maxProfit;
}

int main() {
    vector<int> p1 = {7, 1, 5, 3, 6, 4};
    assert(maxProfit(p1) == 5);

    vector<int> p2 = {7, 6, 4, 3, 1};
    assert(maxProfit(p2) == 0);

    cout << "All tests passed!" << endl;
    return 0;
}
